### PR TITLE
Feature 68 add resource to organization

### PR DESCRIPTION
**Pull request by ignatIgnatov**
I had to make general changes to the mapper, turning it into a custom mapper. This was necessary because of the entity fetcher, events annotations and the integration tests.

### CODE REVIEW

File: src/main/java/jewellery/inventory/dto/request/ResourcePurchaseRequestDto.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/dto/request/ResourcePurchaseRequestDto.java:1]: Obsolete change. To remove:
`git checkout origin/test src/main/java/jewellery/inventory/dto/request/ResourcePurchaseRequestDto.java`

File: src/main/java/jewellery/inventory/mapper/ResourceInOrganizationMapper.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+import org.springframework.stereotype.Component;
+
+@Component
+@RequiredArgsConstructor
+public class ResourceInOrganizationMapper {
+
+  private final OrganizationMapper organizationMapper;
+  private final ResourceMapper resourceMapper;
+
+  public ResourcesInOrganizationResponseDto toResourceInOrganizationResponse(
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/mapper/ResourceInOrganizationMapper.java]:
```suggestion
  public ResourcesInOrganizationResponseDto toResourcesInOrganizationResponse(
```

File: src/main/java/jewellery/inventory/mapper/ResourceInOrganizationMapper.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+      ResourceInOrganization resourceInOrganization) {
+    return ResourcesInOrganizationResponseDto.builder()
+        .owner(organizationMapper.toResponse(resourceInOrganization.getOrganization()))
+        .dealPrice(resourceInOrganization.getDealPrice())
+        .resourcesAndQuantities(List.of(getResourceQuantityResponseDto(resourceInOrganization)))
+        .build();
+  }
+
+  public ResourcesInOrganizationResponseDto toResourceInOrganizationResponse(
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/mapper/ResourceInOrganizationMapper.java]:
```suggestion
  public ResourcesInOrganizationResponseDto toResourcesInOrganizationResponse(
```

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+  private static final Logger logger = LogManager.getLogger(ResourceInOrganizationService.class);
+  private final ResourceInOrganizationRepository resourceInOrganizationRepository;
+  private final UserInOrganizationService userInOrganizationService;
+  private final OrganizationService organizationService;
+  private final ResourceService resourceService;
+  private final ResourceInOrganizationMapper resourceInOrganizationMapper;
+
+  private static final BigDecimal EPSILON = new BigDecimal("1e-10");
+
+  @LogCreateEvent(eventType = EventType.ORGANIZATION_ADD_RESOURCE_QUANTITY)
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: I think this should be a  LogUpdateEvent because we want to see a before and after for the given quantity. You can use ResourceInUserService::addResourceToUser as an "inspiration"

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+
+  public ResourcesInOrganizationResponseDto getAllResourcesFromOrganization(UUID organizationId) {
+    Organization organization = organizationService.getOrganization(organizationId);
+    userInOrganizationService.validateUserInOrganization(organization);
+
+    return resourceInOrganizationMapper.toResourceInOrganizationResponse(organization);
+  }
+
+  @Override
+  public Object fetchEntity(Object... ids) {
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: I know this is a public method but we are keeping the other fetchEntity methods at the bottom of the files. I suggest we move it there for consistency

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+    return value.abs().compareTo(EPSILON) < 0;
+  }
+
+  private boolean isNegative(BigDecimal value) {
+    return value.compareTo(BigDecimal.ZERO) < 0;
+  }
+
+  private Optional<ResourceInOrganization> findResourceInOrganization(
+      Organization organization, UUID resourceId) {
+    logger.info(
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: Should this be a debug log?

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+  private ResourcesInOrganizationResponseDto getResourceInOrganizationResponse(
+      UUID organizationId, UUID resourceId) {
+    Organization organization = organizationService.getOrganization(organizationId);
+    Resource resource = resourceService.getResourceById(resourceId);
+    ResourceInOrganization resourceInOrganization =
+        getResourceInOrganization(organization, resource);
+    if (resourceInOrganization != null) {
+      return resourceInOrganizationMapper.toResourceInOrganizationResponse(resourceInOrganization);
+    }
+    return null;
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: From what I saw getResourceInOrganization cannot return null. It either finds it in the db or it creates it. Am I missing a scenario here?

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+  }
+
+  private ResourcesInOrganizationResponseDto removeQuantityFromResourceNoLog(
+      UUID organizationId, UUID resourceId, BigDecimal quantity) {
+    Organization organization = organizationService.getOrganization(organizationId);
+
+    throwExceptionWhenQuantityToRemoveIsNegative(quantity);
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: This seems obsolete as we have a @ Positive annotation on the quantity in ResourceInOrganizationRequest. Is it needed? Are we using the method in any other way that it doesn't take its arguments from the controller?

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+
+    userInOrganizationService.validateCurrentUserPermission(
+        organization, OrganizationPermission.REMOVE_RESOURCE_QUANTITY);
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: This method might make most sense in the OrganizationService because we are going to check for user permissions everywhere oganization related.

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+
+    ResourceInOrganization resourceInOrganization =
+        findResourceInOrganizationOrThrow(organization, resourceId);
+    removeQuantityFromResource(resourceInOrganization, quantity);
+
+    if (resourceInOrganization != null) {
+      return resourceInOrganizationMapper.toResourceInOrganizationResponse(resourceInOrganization);
+    }
+    return new ResourcesInOrganizationResponseDto();
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: I think we should move the null check handling in the mapper. It makes sense the mapper is able to handle null objects. Then this becomes
```suggestion

    return resourceInOrganizationMapper.toResourceInOrganizationResponse(resourceInOrganization);
 
```

File: src/main/java/jewellery/inventory/service/ResourceInUserService.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceInUserService.java:1]: Obsolete changes. To remove do:
`git checkout origin/test src/main/java/jewellery/inventory/service/ResourceInUserService.java`

File: src/main/java/jewellery/inventory/service/UserInOrganizationService.java
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/UserInOrganizationService.java:1]: Should we move those to Organization? Or are we not using them there. If that's the case, leave it here.

File: src/main/resources/db/migration/dev/V1_11__delete_deal_price_from_resource_in_organization.sql
Comment (VladoKat) [inline src/main/resources/db/migration/dev/V1_11__delete_deal_price_from_resource_in_organization.sql:1]: good! But we need to add this for prod as well. 

We should consider having a base location for the migrations that need to run on every environment and leave the env specific ones to the dev/prod folders. Let's discuss how to do that (not part of this ticket)

File: src/test/java/jewellery/inventory/integration/ResourceInOrganizationCrudIntegrationTest.java [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+    ResourceResponseDto resourceResponse = createResourceResponse();
+
+    ResourceInOrganizationRequestDto request =
+        ResourceInOrganizationTestHelper.createResourceInOrganizationRequestDto(
+            organizationResponseDto.getId(),
+            resourceResponse.getId(),
+            RESOURCE_QUANTITY_TO_REMOVE.subtract(getBigDecimal("1")),
+            RESOURCE_PRICE);
+
+    sendResourceToOrganization(request);
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/ResourceInOrganizationCrudIntegrationTest.java]: group those together as a preparation

File: src/test/java/jewellery/inventory/integration/ResourceInOrganizationCrudIntegrationTest.java [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+
+    ResponseEntity<ResourcesInOrganizationResponseDto> response =
+        sendDeleteOperation(getDeleteResourceUrl(organizationResponseDto, resourceResponse));
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/ResourceInOrganizationCrudIntegrationTest.java:248]: Have this separated both above and below it so that it is separated as "action" and the the last separation is assertions

File: src/test/java/jewellery/inventory/unit/mapper/ResourceInOrganizationMapperTest.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/unit/mapper/ResourceInOrganizationMapperTest.java:1]: Tests look good! I would only do the separations of the code with 3 groups separated with a single new line like this:
```
test(){
  <preparation>

  <action>
  
  <assertions>
}
```

File: src/test/java/jewellery/inventory/unit/service/ResourceInOrganizationServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+
+    Assertions.assertThrows(
+        MissingOrganizationPermissionException.class,
+        () ->
+            resourceInOrganizationService.addResourceToOrganization(
+                resourceInOrganizationRequestDto));
+  }
+
+  @Test
+  void testAddResourceToOrganizationShouldThrowResourceNotFoundException() {
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/unit/service/ResourceInOrganizationServiceTest.java]: I think we are duplicating this in the integration test

File: src/test/java/jewellery/inventory/unit/service/ResourceInOrganizationServiceTest.java [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+        .validateCurrentUserPermission(organization, OrganizationPermission.ADD_RESOURCE_QUANTITY);
+    verify(resourceService, times(1))
+        .getResourceById(resourceInOrganizationRequestDto.getResourceId());
+    verify(resourceInOrganizationMapper, times(1))
+        .toResourceInOrganizationResponse(resourceInOrganization);
+    verify(resourceInOrganizationRepository, times(1)).save(resourceInOrganization);
+  }
+
+  @Test
+  void testRemoveQuantityFromResourceShouldThrowOrganizationNotFoundException() {
```
Comment (VladoKat) [inline src/test/java/jewellery/inventory/unit/service/ResourceInOrganizationServiceTest.java:110]: Same with this

File: src/test/java/jewellery/inventory/unit/service/ResourceInOrganizationServiceTest.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/unit/service/ResourceInOrganizationServiceTest.java:1]: Tests look good. Let's not duplicate them though. If we have tested something with integration test, there is no need to test it again with unit test. Let me know if we are testing something specific that isn't exactly covered in the integration "bad case" tests.

**Review by VladoKat — CHANGES_REQUESTED** [review]
PR looks good. Minor tweaks required here and there. Let me know what you think

File: src/main/java/jewellery/inventory/dto/request/ResourcePurchaseRequestDto.java
Comment (ignatIgnatov) [reply src/main/java/jewellery/inventory/dto/request/ResourcePurchaseRequestDto.java:1]: Done!

File: src/main/java/jewellery/inventory/mapper/ResourceInOrganizationMapper.java (hunk 1 above)
Comment (ignatIgnatov) [reply src/main/java/jewellery/inventory/mapper/ResourceInOrganizationMapper.java]: Done!

File: src/main/java/jewellery/inventory/mapper/ResourceInOrganizationMapper.java (hunk 2 above)
Comment (ignatIgnatov) [reply src/main/java/jewellery/inventory/mapper/ResourceInOrganizationMapper.java]: Done!

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java (hunk 4 above)
Comment (ignatIgnatov) [reply src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: Done!

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java (hunk 5 above)
Comment (ignatIgnatov) [reply src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: Done!

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java (hunk 6 above)
Comment (ignatIgnatov) [reply src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: Done!

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java (hunk 3 above)
Comment (ignatIgnatov) [reply src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: Done!

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java (hunk 7 above)
Comment (ignatIgnatov) [reply src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: Done! After the fix validation comes from the controller

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java (hunk 8 above)
Comment (ignatIgnatov) [reply src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: Done! I moved them to Organization service and refactored the code and some tests

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java (hunk 9 above)
Comment (ignatIgnatov) [reply src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: Done! I moved this check to the mapper

File: src/main/java/jewellery/inventory/service/ResourceInUserService.java
Comment (ignatIgnatov) [reply src/main/java/jewellery/inventory/service/ResourceInUserService.java:1]: Done!

File: src/main/resources/db/migration/dev/V1_11__delete_deal_price_from_resource_in_organization.sql
Comment (ignatIgnatov) [reply src/main/resources/db/migration/dev/V1_11__delete_deal_price_from_resource_in_organization.sql:1]: I forgot to put it there too. It's there now.

Comment (ignatIgnatov) [reply src/main/resources/db/migration/dev/V1_11__delete_deal_price_from_resource_in_organization.sql:1]: I am open to discuss this

File: src/test/java/jewellery/inventory/unit/service/ResourceInOrganizationServiceTest.java (hunk 12 above)
Comment (ignatIgnatov) [reply src/test/java/jewellery/inventory/unit/service/ResourceInOrganizationServiceTest.java]: Done!

File: src/test/java/jewellery/inventory/unit/service/ResourceInOrganizationServiceTest.java (hunk 13 above)
Comment (ignatIgnatov) [reply src/test/java/jewellery/inventory/unit/service/ResourceInOrganizationServiceTest.java:110]: Done!

File: src/test/java/jewellery/inventory/integration/ResourceInOrganizationCrudIntegrationTest.java (hunk 10 above)
Comment (ignatIgnatov) [reply src/test/java/jewellery/inventory/integration/ResourceInOrganizationCrudIntegrationTest.java]: Done!

File: src/test/java/jewellery/inventory/integration/ResourceInOrganizationCrudIntegrationTest.java (hunk 11 above)
Comment (ignatIgnatov) [reply src/test/java/jewellery/inventory/integration/ResourceInOrganizationCrudIntegrationTest.java:248]: Done!

File: src/main/java/jewellery/inventory/mapper/ResourceInOrganizationMapper.java [hunk 14]
```diff
@@ -0,0 +1,52 @@
...
+  public ResourcesInOrganizationResponseDto toResourcesInOrganizationResponse(
+      ResourceInOrganization resourceInOrganization) {
+    if (resourceInOrganization != null) {
+      return ResourcesInOrganizationResponseDto.builder()
+          .owner(organizationMapper.toResponse(resourceInOrganization.getOrganization()))
+          .dealPrice(resourceInOrganization.getDealPrice())
+          .resourcesAndQuantities(List.of(getResourceQuantityResponseDto(resourceInOrganization)))
+          .build();
+    }
+    return new ResourcesInOrganizationResponseDto();
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/mapper/ResourceInOrganizationMapper.java]: I would think this should return null if the input is null. Do we have some NullPointerException thrown if it is null?

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java [hunk 15]
```diff
@@ -0,0 +1,199 @@
...
+  }
+
+  private ResourcesInOrganizationResponseDto getResourceInOrganizationResponse(
+      UUID organizationId, UUID resourceId) {
+    ResourceInOrganization resourceInOrganization =
+        getResourceInOrganization(organizationId, resourceId);
+    if (resourceInOrganization != null) {
+      return resourceInOrganizationMapper.toResourcesInOrganizationResponse(resourceInOrganization);
+    }
+    return null;
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: This check shouldn't be needed if we've moved the null check in the mapper.

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java [hunk 16]
```diff
@@ -0,0 +1,199 @@
...
+
+        return getResourceInOrganizationResponse(
+            resourceInOrganizationRequestDto.getOrganizationId(),
+            resourceInOrganizationRequestDto.getResourceId());
+      } else {
+        return getResourceInOrganizationResponse((UUID) ids[0], (UUID) ids[1]);
+      }
+    }
+    return null;
+  }
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java:195]: Let's move it to the bottom

File: src/main/java/jewellery/inventory/service/UserInOrganizationService.java [hunk 17]
```diff
@@ -39,13 +38,19 @@ public OrganizationMembersResponseDto getAllUsersInOrganization(UUID organizatio
...
+    organizationService.validateCurrentUserPermission(
         organizationService.getOrganization(organizationId), OrganizationPermission.MANAGE_USERS);
 
     UserInOrganization userInOrganization =
         getUserInOrganizationByUserIdAndOrganizationId(userId, organizationId);
 
+
+    Organization organization = userInOrganization.getOrganization();
+
+    organizationService.validateCurrentUserPermission(organization, OrganizationPermission.MANAGE_USERS);
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/UserInOrganizationService.java]: Checking userPermissions twice?

File: src/main/java/jewellery/inventory/service/UserInOrganizationService.java [hunk 18]
```diff
@@ -90,20 +97,7 @@ public void deleteUserInOrganization(UUID userId, UUID organizationId) {
...
-          currentUser.getId(), organization.getId(), permission);
-    }
-    logger.debug(
-        "User permission validation successful. User ID: {}, Organization ID: {}",
-        currentUser.getId(),
-        organization.getId());
-  }
-
-  private void validateUserInOrganization(Organization organization) {
+  public void validateUserInOrganization(Organization organization) {
```
Comment (VladoKat) [inline src/main/java/jewellery/inventory/service/UserInOrganizationService.java]: Maybe we should move this to OrganizationService as well. Now that I think about it, it might make sense we keep both `validateUserInOrganization` and `validateCurrentUserPermission` to UserInOrganizationService. But regardless of where, i feel these should be kept in one place.

File: src/test/java/jewellery/inventory/unit/mapper/ResourceInOrganizationMapperTest.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/unit/mapper/ResourceInOrganizationMapperTest.java:1]: Now that we have a null check we should add a unit test for the null case

**Review by VladoKat — CHANGES_REQUESTED** [review]

File: src/main/java/jewellery/inventory/mapper/ResourceInOrganizationMapper.java (hunk 14 above)
Comment (LuboKar) [reply src/main/java/jewellery/inventory/mapper/ResourceInOrganizationMapper.java]: Done!

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java (hunk 15 above)
Comment (LuboKar) [reply src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java]: Done!

File: src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java (hunk 16 above)
Comment (LuboKar) [reply src/main/java/jewellery/inventory/service/ResourceInOrganizationService.java:195]: Done!

File: src/main/java/jewellery/inventory/service/UserInOrganizationService.java (hunk 17 above)
Comment (LuboKar) [reply src/main/java/jewellery/inventory/service/UserInOrganizationService.java]: Done!

File: src/main/java/jewellery/inventory/service/UserInOrganizationService.java (hunk 18 above)
Comment (LuboKar) [reply src/main/java/jewellery/inventory/service/UserInOrganizationService.java]: Done!

File: src/test/java/jewellery/inventory/unit/mapper/ResourceInOrganizationMapperTest.java
Comment (LuboKar) [reply src/test/java/jewellery/inventory/unit/mapper/ResourceInOrganizationMapperTest.java:1]: Done!

File: src/test/java/jewellery/inventory/integration/ProductCrudIntegrationTest.java
Comment (VladoKat) [inline src/test/java/jewellery/inventory/integration/ProductCrudIntegrationTest.java:1]: Obsolete change. Let's not keep it in.
`git checkout origin/test src/test/java/jewellery/inventory/integration/ProductCrudIntegrationTest.java`

**Review by VladoKat — CHANGES_REQUESTED** [review]
Last minor change

File: src/test/java/jewellery/inventory/integration/ProductCrudIntegrationTest.java
Comment (LuboKar) [reply src/test/java/jewellery/inventory/integration/ProductCrudIntegrationTest.java:1]: Done!

**Review by VladoKat — APPROVED** [review]
